src/transaction/input/scripthash.js

```diff
@@ -0,0 +1,148 @@
+'use strict';
+
+var _ = require('lodash');
+var inherits = require('inherits');
+var Input = require('./input');
+var Output = require('../output');
+var $ = require('../../util/preconditions');
+
+var Script = require('../../script');
+var Signature = require('../../crypto/signature');
+var Sighash = require('../sighash');
+var PublicKey = require('../../publickey');
+var BufferUtil = require('../../util/buffer');
+var TransactionSignature = require('../signature');
+
+/**
+ * @constructor
+ */
+function ScriptHashInput(input, pubkeys, redeemScript) {
+  Input.apply(this, arguments);
+  var self = this;
+  this.publicKeys = pubkeys || input.publicKeys;
+  this.threshold = 1;
+  this.redeemScript = redeemScript;
+  $.checkState(Script.buildScriptHashOut(this.redeemScript).equals(this.output.script),
+    'Provided redeemScript doesn\'t hash to the provided output');
+  this.publicKeyIndex = {};
+  this.publicKeys.forEach((publicKey, index) => {
+    self.publicKeyIndex[publicKey.toString()] = index;
+  });
+  // Empty array of signatures
+  this.signatures = new Array(this.publicKeys.length);
+}
+inherits(ScriptHashInput, Input);
+
+ScriptHashInput.prototype.toObject = function() {
+  var obj = Input.prototype.toObject.apply(this, arguments);
+  obj.threshold = this.threshold;
+  obj.publicKeys = this.publicKeys.map(publicKey => publicKey.toString());
+  obj.signatures = this._serializeSignatures();
+  return obj;
+};
+
+ScriptHashInput.prototype._deserializeSignatures = function(signatures) {
+  return signatures.map(signature => signature ? new TransactionSignature(signature) : undefined);
+};
+
+ScriptHashInput.prototype._serializeSignatures = function() {
+  return this.signatures.map(signature => signature ? signature.toObject() : undefined);
+};
+
+ScriptHashInput.prototype.getSignatures = function(transaction, privateKey, index, sigtype) {
+  $.checkState(this.output instanceof Output, 'Malformed output found when signing transaction');
+  sigtype = sigtype || (Signature.SIGHASH_ALL |  Signature.SIGHASH_FORKID);
+
+  var self = this;
+  var results = [];
```

Comment:
Above we converted a similar hunk into something a bit more functionally oriented (filter(), ...). Should we use the same pattern here?